### PR TITLE
design-proposal: instancetype.kubevirt.io - pkg/instancetype refactor

### PR DESCRIPTION
**Pull request by lyarwood**
**What this PR does / why we need it**:

This isn't required at all but I wanted to write down how I think we should go about refactoring the package.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by lyarwood** [comment]
/cc @0xFelix 
/cc @akrejcir 

If you have time this week would you mind giving this a once over? Writing this effort up as a design proposal is a little over the top but I thought it would be easier to discuss now instead of later in a huge PR reworking everything.

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+* All existing functionality should be unchanged by this work, all functional tests should therefore be untouched
+
+## Non Goals
+
+* Any behavioral changes to the API during the release v1.4 cycle should happen outside of this work
+
+## User Stories
+
+* As a developer working on `pkg/instancetype` I want to see the code logically separated for easier maintenance and future improvements  
+* As a developer working on `pkg/instancetype` I want to see the API graduate to `v1`
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:23]: It is not part of this proposal though?

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+* Replace the public `InstancetypesMethods` interface with private smaller interfaces defined closer to their consumers
+* All existing functionality should be unchanged by this work, all functional tests should therefore be untouched
+
+## Non Goals
+
+* Any behavioral changes to the API during the release v1.4 cycle should happen outside of this work
+
+## User Stories
+
+* As a developer working on `pkg/instancetype` I want to see the code logically separated for easier maintenance and future improvements
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:22]:
```suggestion
* As a developer working on instance types and preferences I want to see the code logically separated for easier maintenance and future improvements
```

When done we will have `pkg/instancetype`, `pkg/preference`, etc.?

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+```go
+func (c *VMController) Sync(){
+    [..]
+    spec, err := c.instancetypeHandler.Find(vm)
+    [..]
+}
+
+```
+
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:147]: Bonus: No more stubs required in tests.

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md (hunk 1 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:23]: In my mind this refactor is something we need to do ahead of the API hitting v1 but I hadn't made that explicit anywhere else in the doc. I'll add something to goals etc.

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md (hunk 3 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:147]: I don't see how this changes anything tbh, we can still switch out the implementations in unit tests with stubs if we want to.

Comment (0xFelix) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:147]: I meant you are no longer required to implement stubs for interface methods you are not interested in when using reduced private interfaces.

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        Store:                  instancetypeStore,
+        ClusterStore:           clusterInstancetypeStore,
+        ControllerRevisionStore controllerRevisionStore,
+        Client:                 virtClient,
+    },
+}
+
+c := &VMController{
+    [..]
+    instancetypeHandler: instancetypeHandler
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:132]: The idea is to decouple the implementation from the expected behavior.

See

https://github.com/kubevirt/kubevirt/blob/493ae9dc42ed6c83063579b6114968790ab4fd82/docs/coding-conventions.md?plain=1#L237

[the remark this answers is not in the document]
Comment (akrejcir) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:132]: Ok, I understand. I haven't seen this usage of interfaces in the codebases much.

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+As set out above the implementation of this logic will reside under `pkg/instancetype/find` and will satisfy this interface.
+
+```go
+type Finder struct {
+    Store                   cache.Store
+    ClusterStore            cache.Store
+    ControllerRevisionStore cache.Store
+    Client                  kubevirtcli.KubevirtClient
+}
```
[the remark this answers is not in the document]
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:99]: Yup that's a mistake, the fields should be private. I'll update.

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+The logic around applying instance types and preferences is significantly different and should be clearly separated.
+
+Further separation should also allow for specific preferences to be applied in specific use cases.
+
+An example being during device hot-plug where a subset of preferences need to be applied, for example network interface or disk preferences.
+
+## Move infer code under `pkg/{instancetype,preference}/infer`
+
+Any common code placed under `pkg/instancetype`.
```
[the remark this answers is not in the document]
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:51]: Yup apologies this should actually say `pkg/instancetype/infer` but I copy pasted between sections, I'll update.

File: design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md:132]: Yeah the networking folks are using it more and more so I think it's a valid pattern to use now given the coding convention suggestions.

**Comment by akrejcir** [comment]
Looks good, thanks for the changes.

**Review by jean-edouard — COMMENTED** [review]
/approve

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Pull request merged**